Refactor DB as document store, add Postgres, remove mongo

eelcovdw opened the pull request:
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

eelcovdw committed:
Merge branch 'eelco/json_serde' into aziz/sql_json

eelcovdw committed:
use new json serde, add enum support

eelcovdw committed:
usercode, status, log

eelcovdw committed:
organize files

eelcovdw committed:
cleaning

eelcovdw committed:
add attr_searchable

eelcovdw committed:
add order_by for sync stash

eelcovdw committed:
sync stash

abyesilyurt committed:
implement code history

eelcovdw committed:
blob store, contains

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
key to str

eelcovdw committed:
update permissions

eelcovdw committed:
action store WIP

abyesilyurt committed:
Merge remote-tracking branch 'origin/aziz/sql_json' into eelco/json_action_store

abyesilyurt committed:
implement action stash

abyesilyurt committed:
implement storage permissions

abyesilyurt committed:
action object fixes

abyesilyurt committed:
fix get_mock

eelcovdw committed:
add exclude attrs

abyesilyurt committed:
Merge branch 'eelco/json_action_store' of github.com:OpenMined/PySyft into eelco/json_action_store

eelcovdw committed:
fix deserialize

abyesilyurt committed:
Merge branch 'eelco/json_action_store' of github.com:OpenMined/PySyft into eelco/json_action_store

abyesilyurt committed:
fix error handling

abyesilyurt committed:
type hint fixes

abyesilyurt committed:
fix perms

abyesilyurt committed:
fix test

abyesilyurt committed:
fix get_perms

abyesilyurt committed:
fix sync

kiendang committed:
Add postgres document store

kiendang committed:
Add postgres action store

abyesilyurt committed:
add project stash

abyesilyurt committed:
Merge remote-tracking branch 'origin/dev' into aziz/sql_json

khoaguin committed:
Merge branch 'dev' into postgres

khoaguin committed:
[syft/stores] Update `server.py` to use `PostgreSQL` stores

eelcovdw committed:
move sqlite connection to Server

eelcovdw committed:
move sqlite connection to Server

abyesilyurt committed:
implement notification stash

abyesilyurt committed:
signature handling improvements

khoaguin committed:
Merge branch 'dev' into postgres

khoaguin committed:
[syft/server] replace mongo store config with postgres
- add environment and settings variables for postgres
- new error handling for postgresql store

eelcovdw committed:
merge WIP

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

eelcovdw committed:
add stash

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
fix bugs

eelcovdw committed:
fix first notebook

eelcovdw committed:
fix settingsstash

abyesilyurt committed:
fix order

eelcovdw committed:
fix sorting, add get_index

eelcovdw committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

khoaguin committed:
Merge branch 'dev' into postgres

abyesilyurt committed:
fix bugs

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
bug fix

madhavajay committed:
Merge branch 'dev' into postgres

abyesilyurt committed:
add basic postgres support

madhavajay committed:
Made some progress on postgreql

abyesilyurt committed:
fix sync

abyesilyurt committed:
Merge remote-tracking branch 'origin/aziz/fix_sync' into aziz/postgres_json

shubham3121 committed:
Merge branch 'dev' into postgres

shubham3121 committed:
Add helm templates for postgres
- add postgres service to devspace
- pass postgres creds to backend

Co-authored-by: khoaguin <dkn.work@protonmail.com>

khoaguin committed:
Merge branch 'dev' into postgres

khoaguin committed:
[syft/stores] new exception handlings for some methods in `SQLiteBackingStore`
[tox] trying to delete a datasite cluster before launching

khoaguin committed:
[syft/store] fix wrong imports and ports for postgres store

eelcovdw committed:
merge dev

eelcovdw committed:
fix

abyesilyurt committed:
wip

abyesilyurt committed:
Merge remote-tracking branch 'origin' into aziz/postgres_json

abyesilyurt committed:
Merge branch 'aziz/sql_json' into aziz/postgres_json

eelcovdw committed:
Merge branch 'dev' into aziz/sql_json

abyesilyurt committed:
remove dict document store

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
fix settings stash test

eelcovdw committed:
add remaining stashes

eelcovdw committed:
fix

eelcovdw committed:
fixes

abyesilyurt committed:
fix notifcation tests

abyesilyurt committed:
fix request stash tests

abyesilyurt committed:
fix user stash tests

abyesilyurt committed:
fix dataset stash tests

abyesilyurt committed:
fix base stash tests

eelcovdw committed:
json serialize filter

eelcovdw committed:
fix

eelcovdw committed:
Merge pull request #9263 from OpenMined/eelco/remaining_stashes

[wip] rewrite remaining stashes

abyesilyurt committed:
fix notification service tests

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
fix request service

eelcovdw committed:
fix protocol

eelcovdw committed:
merge dev

eelcovdw committed:
remove dataset stash any type

eelcovdw committed:
fix dependencies

abyesilyurt committed:
fix project tests

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

eelcovdw committed:
fix job unittests

eelcovdw committed:
merge

abyesilyurt committed:
simple way to create stashes

eelcovdw committed:
fix queue

eelcovdw committed:
fix jobservice

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
make traceback clickable

koenvanderveen commented on packages/syft/src/syft/service/notification/notification_stash.py:
```diff
@@ -102,14 +69,10 @@ def get_notification_for_linked_obj(
-        return self.query_one(credentials=credentials, qks=qks).unwrap(
-            public_message=f"Notifications for Linked Object {linked_obj} not found"
-        )
+        # TODO does this work?
```
can we not just match on the id?

abyesilyurt committed:
postgres bug fixes

khoaguin committed:
Merge branch 'dev' into postgres

abyesilyurt committed:
wip postgres integration

abyesilyurt committed:
comment out postgres

khoaguin committed:
[k8s] stop creating the mongo pod in the cluster
[chores] improve some debug printings

khoaguin committed:
Merge branch 'dev' into postgres

khoaguin committed:
[syft/postgre_store] fix store initialization by adding `CREATE TABLE IF NOT EXISTS`

khoaguin committed:
[lint] fix some linting issues

khoaguin committed:
[tox] waiting for postgres instead of mongo service in k8s tests

khoaguin committed:
[notebooks] revert api notebook `0-load-data` back to `dev`'s state

khoaguin committed:
[postgres] add error handling for getting postgres storage config from json

abyesilyurt committed:
Merge remote-tracking branch 'origin/dev' into aziz/sql_json

eelcovdw committed:
fix consumer

khoaguin committed:
[stores] refactor `_execute` to be a static method for both sqlite and postgre backing store
- postgre's cursor is not cached and renewed whenever a command is executed

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

eelcovdw committed:
add Query, split table creation

abyesilyurt committed:
fix queue stash tests

eelcovdw committed:
remove old methods

abyesilyurt committed:
fix action store tests

abyesilyurt committed:
fix worker pool tests

abyesilyurt committed:
fix some tests

abyesilyurt committed:
Merge remote-tracking branch 'origin/dev' into aziz/sql_json

khoaguin committed:
Merge branch 'dev' into postgres

shubham3121 committed:
scope the use of cursor and db connection

shubham3121 committed:
fix lint

shubham3121 committed:
fix _delete args in postgres sqlstore

abyesilyurt committed:
fix iter and order

eelcovdw committed:
WIP migrationservice

eelcovdw committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

eelcovdw committed:
merge

eelcovdw committed:
fix migrations

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
fix test

eelcovdw committed:
fix test

eelcovdw committed:
merge

eelcovdw committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

shubham3121 committed:
add a few log statements

eelcovdw committed:
merge

shubham3121 committed:
set prepare_threshold to None

eelcovdw committed:
fix new query

abyesilyurt committed:
fix more tests

eelcovdw committed:
Merge pull request #9277 from OpenMined/eelco/split-objectstash

[WIP] split db logic from new json stash

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
fix all unit tests

eelcovdw committed:
refactor get_one, add validation to limit offset

eelcovdw committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

eelcovdw committed:
fix dataset get_all

abyesilyurt committed:
implement reset password flow for python and fix it

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
fix some linting errors

eelcovdw committed:
fix is_unique, get_role

eelcovdw committed:
merge

eelcovdw committed:
comments

eelcovdw committed:
fix set type annotation

eelcovdw committed:
local session management

khoaguin committed:
Merge branch 'dev' into postgres

khoaguin committed:
[clean] remove mongo related tests and fixtures

khoaguin committed:
[clean] remove mongo from tox.ini and helm's `base.yaml`
- scanning for for postgres container security instead of mongo

khoaguin committed:
[clean] delete mongo code from document store
- remove mongo from gcp helm charts

abyesilyurt committed:
Merge remote-tracking branch 'origin' into aziz/sql_json

khoaguin committed:
[clean] remove mongomock in tests and mongo-related packages

khoaguin committed:
[clean] remove mongo related api reference
- remove ununsed scripts

abyesilyurt committed:
fix lint

khoaguin committed:
[script] `reset_k8s.sh` works with postgres

eelcovdw committed:
fix session

abyesilyurt committed:
replace DocumentStore with DBManager

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

eelcovdw committed:
fix env variables in module scope, fix root admin already exists

eelcovdw committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

khoaguin committed:
Merge branch 'dev' into postgres

shubham3121 committed:
add a connection pool class
add a query executor class

khoaguin committed:
Merge branch 'dev' into postgres

shubham3121 committed:
Merge branch 'aziz/sql_json' into shubham/postgres-k8s

shubham3121 committed:
Merge branch 'dev' into aziz/sql_json

shubham3121 committed:
Merge branch 'aziz/sql_json' into shubham/postgres-k8s

eelcovdw committed:
add protocol

eelcovdw committed:
split db manager class

eelcovdw committed:
rename db_sqlite

abyesilyurt committed:
add partial updates

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

eelcovdw committed:
root admin delete check

eelcovdw committed:
fix

eelcovdw committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

eelcovdw committed:
fix tests that delete root user

abyesilyurt committed:
implement nested queries

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

shubham3121 committed:
remove duplicate postgres reference in base.yml
move db manager and db config to base.py
define postgres db manager
fix json_quote based on postgres or sqlite
fix lint

abyesilyurt committed:
add a test for todo item

abyesilyurt committed:
clean up todo notes

shubham3121 committed:
rename postgres_db to postgres.py

shubham3121 committed:
rename base to sqlite_db.py
rename sqilte_db.py to sqlite.py

eelcovdw committed:
Merge branch 'dev' into aziz/sql_json

shubham3121 committed:
Merge branch 'aziz/sql_json' into shubham/postgres-k8s

shubham3121 committed:
Merge branch 'aziz/sql_json' into shubham/postgres-k8s

shubham3121 committed:
fix imports

Co-authored-by: khoaguin <dkn.work@protonmail.com>

shubham3121 committed:
update postgres version to 16.1
update init_stores to return db manager

khoaguin committed:
[deps] scanning postgres:16.1 security

eelcovdw committed:
fixes

abyesilyurt committed:
fix network service

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
Merge branch 'shubham/postgres-k8s' into aziz/sql_json

abyesilyurt committed:
update protocol file

abyesilyurt committed:
add db config

eelcovdw committed:
fix migration tests

eelcovdw committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

khoaguin committed:
- [db] fix failed stack.notebook test since by changing `SQLITE_PATH` to /tmp/syft instead of `HOME/data/db/`
- [dockerfile] fix `WARN: FromAsCasing: 'as' and 'FROM' keywords' casing do not match`
- [deps] update `pyarrow` version

khoaguin committed:
Merge branch 'dev' into aziz/sql_json

khoaguin commented on packages/grid/backend/grid/core/config.py:
```diff
@@ -138,7 +138,7 @@ def get_emails_enabled(self) -> Self:
     )
     N_CONSUMERS: int = int(os.getenv("N_CONSUMERS", 1))
-    SQLITE_PATH: str = os.path.expandvars("$HOME/data/db/")
+    SQLITE_PATH: str = os.path.expandvars("/tmp/syft/")
```
Hi @shubham3121 I changed this to `/tmp/syft` to solve sqlalchemy can't open database file in `/root/data/db`, but not sure it's optimal. Also in `postgres-statefulset.yaml`  the `mountPath` is set to `/data/db`, should be change it to be the same to value of `SQLITE_PATH`?

shubham3121 replied:
What error do we get with `/root/data/db` ? Does it complain about directory doesn't exists or some sort of permission error ? Earlier we used to dir if doesn't exist, I think sqlalchemy could do that same maybe via some config.

khoaguin replied:
The error is like below. I think it's a permission issue. This made the `stack.test.notebook` failed since the backend container was not running

![Screenshot from 2024-09-13 10-55-57](https://github.com/user-attachments/assets/611b909a-f420-40e9-8256-5fb65b71c24b)

shubham3121 replied:
Its seems like a permission issue to create a directory at the root level. Maybe we can create the director beforehand (while changing ownership of directory to user)

khoaguin replied:
but is there any problem with using `/tmp/syft` since we use it in the local python version?

khoaguin replied:
after discussion changed it to `/tmp/data/db` for both single container and k8s mode in this commit [cc8e976](https://github.com/OpenMined/PySyft/pull/9176/commits/cc8e9762d75e300ce522daf007b40db65c4eacee)

abyesilyurt committed:
add different base to handle two dbs

eelcovdw committed:
add get_table to basecls

abyesilyurt committed:
fix postgres error

abyesilyurt committed:
fix queries for postgres

abyesilyurt committed:
Merge branches 'aziz/sql_json' and 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
fix sqlite

abyesilyurt committed:
fix ordering in postgres

abyesilyurt committed:
fix string casting in postgres

eelcovdw committed:
deduplicate permissions

eelcovdw committed:
merge

eelcovdw committed:
fix getlatest

eelcovdw committed:
fix orderby non-json

eelcovdw committed:
fix lint CI

eelcovdw committed:
fix lint

abyesilyurt committed:
remove uses of find_all

eelcovdw committed:
fix instrument imports

abyesilyurt committed:
trying to fix flaky test

shubham3121 committed:
Merge branch 'dev' into aziz/sql_json

shubham3121 commented on packages/syft/src/syft/protocol/protocol_version.json:
```diff
@@ -1,5 +1,1172 @@
 {
-  "1": {
-    "release_name": "0.9.1.json"
```
@eelcovdw @abyesilyurt Its seems like we have reset the protocol file. Do we want to ensure compatibility with previous version ?

khoaguin committed:
[protocol_version] define old version for `SyftWorkerImage` and `Notification`
- revert protocol json files back to dev's state

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

koenvanderveen commented on packages/syft/src/syft/orchestra.py:
```diff
@@ -183,7 +183,9 @@ def deploy_to_python(
     log_level: str | int | None = None,
     debug: bool = False,
     migrate: bool = False,
+    store_client_config: dict | None = None,
```
is this convenient to pass, do we have an example of how we would pass it? Also, how do I know the default value?

khoaguin replied:
it's old code and should be removed

khoaguin replied:
Removed in [3b3023d](https://github.com/OpenMined/PySyft/pull/9176/commits/3b3023d3aa3f3a9914b276a0663f28c4feebe0a0)

koenvanderveen commented on packages/syft/src/syft/serde/json_serde.py:
```diff
@@ -0,0 +1,457 @@
+@dataclass
+class JSONSerde(Generic[T]):
+    klass: type[T]
+    serialize_fn: Callable[[T], Json] | None = None
```
why are these optional if the only usage is that it always has them

koenvanderveen replied:
I think this is a bit clunky to read, can we not just have a type check in the register function that checks if its a default json type and then passes a no-op into this?

shubham3121 committed:
reset queue version to 1 and remove migrations for queue and subclasses
- extract db config from server
- stage protocol version

koenvanderveen commented on packages/syft/src/syft/serde/json_serde.py:
```diff
@@ -0,0 +1,457 @@
+    elif _is_serializable_mapping(annotation):
+        result = _serialize_mapping_to_json(value, annotation)
+    else:
+        result = _serialize_to_json_bytes(value)
```
why do we need this? We also still store the bytes separately I think? would we ever query on the bytes?

eelcovdw replied:
Many types in syft are not serializable to json, either because they are custom types or because they are untypedd (and cannot be serialize back to the python obj). they are stored inside the json as bytestring.

for example: UserCodeStatusCollection is not serializable because the keys are some custom object, so the db would look like:

```
{
    "__canonical_name__": "UserCodeStatusCollection",
    "__version__": 1,
    "status_dict": <bytestring>,
}
```

koenvanderveen replied:
yes, but isnt that redundant?

eelcovdw replied:
There is no other place we save the full blob, the json in Table.fields is the only data we're saving in the db

koenvanderveen commented on packages/syft/src/syft/serde/json_serde.py:
```diff
@@ -0,0 +1,457 @@
+    return None
+
+
+def _serialize_searchable_attrs(
```
we should probably rename this to something that indicates that those are added

koenvanderveen commented on packages/syft/src/syft/serde/json_serde.py:
```diff
@@ -0,0 +1,457 @@
+            try:
+                value = getattr(obj, attr)
+            except Exception as e:
+                if raise_errors:
```
when would you not want to do this?

eelcovdw replied:
searchable attrs can be `@property` methods, and those can throw errors. Example: `Job.code_id` was throwing errors when the Job was for a twinAPI. This flag is to control this behaviour, 'do i want to skip writing if some method throws an error'

we're using raise_errors=False right now to just skip searchable properties that hit an exception

koenvanderveen commented on packages/syft/src/syft/server/server.py:
```diff
@@ -1717,17 +1722,32 @@ def create_initial_settings(self, admin_email: str) -> ServerSettings:
 
 
-def create_admin_new(
+def create_root_admin(
```
we should rename this to reflect the logic

koenvanderveen replied:
it also feels like there is a third option: the root admin exists, with this email

eelcovdw replied:
Then it would fail the first check, root verify key user already exists

shubham3121 committed:
fix lint

shubham3121 committed:
fix arg to refer to canonical name if exists
- redefine migrations for queue items and subclasses

khoaguin committed:
[clean] remove `store_client_config` arg using for debug

shubham3121 committed:
fix migration logic for worker settings
- add migration method for notification class

abyesilyurt committed:
fix partial updates

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

shubham3121 committed:
fix return statement in queue stash migration

eelcovdw committed:
rename create_root_admin

eelcovdw committed:
rename searchable attrs serde

eelcovdw committed:
jsonserde default noop

khoaguin committed:
change sqlite db path in container mode and postgresql mount path in k8s to `tmp/data/db`

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

shubham3121 committed:
fix class name for APIEndpointQueueItem version 1

shubham3121 commented on packages/syft/setup.cfg:
```diff
@@ -111,7 +114,6 @@ telemetry =
     opentelemetry-instrumentation==0.48b0
     opentelemetry-instrumentation-requests==0.48b0
     opentelemetry-instrumentation-fastapi==0.48b0
-    opentelemetry-instrumentation-pymongo==0.48b0
```
We need to add `opentelemetry-instrumentation-sqlalchemy==0.48b0`

shubham3121 replied:
Resolved by: https://github.com/OpenMined/PySyft/pull/9176/commits/7a6458883145d1efd971513921cb83a4e4ca8114

shubham3121 committed:
add sqlalchemy opentelemetry intrumentation package to pypi

abyesilyurt committed:
refactor db reset logic

abyesilyurt committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

abyesilyurt committed:
use in memory sqlite

abyesilyurt committed:
manage db path externally

abyesilyurt committed:
Revert "manage db path externally"

This reverts commit 8417653217fb14883f0baf13d40599acd291b4a5.

eelcovdw committed:
more logs in custom endpoint notebook

eelcovdw committed:
Merge branch 'aziz/sql_json' of github.com:OpenMined/PySyft into aziz/sql_json

eelcovdw committed:
debug partition_by_server

eelcovdw committed:
wrap DatabaseError

eelcovdw committed:
error handling for Query db errors
